src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -4233,15 +4282,365 @@ module Mutations = struct
+            ; snapp_update = Account_update.Update.dummy
+            ; preconditions = None
+            ; authorization_kind = Account_update.Authorization_kind.Signature
+            }
+          in
+          let zkapp_command =
+            Transaction_snark.For_tests.deploy_snapp ~constraint_constants
+              ~default_permissions:true spec
+          in
+          let rec go () =
+            match%bind send_zkapp_command mina zkapp_command with
+            | Ok _ ->
+                fee_payer_nonce := Account.Nonce.succ !fee_payer_nonce ;
+                account_creator_nonce :=
+                  Account.Nonce.succ !account_creator_nonce ;
+                [%log info]
+                  "Successfully send out zkApp $command that creates a zkApp \
+                   account"
+                  ~metadata:
+                    [ ("command", Zkapp_command.to_yojson zkapp_command) ] ;
+                Deferred.unit
+            | Error err ->
+                [%log info] "Failed to setup a zkApp account, try again"
+                  ~metadata:
+                    [ ("zkapp_command", Zkapp_command.to_yojson zkapp_command)
+                    ; ("error", `String err)
+                    ] ;
+                go ()
+          in
+          go () )
+
+    let is_zkapp_deployed kp ledger =
+      match
+        Option.try_with (fun () ->
+            let account = account_of_kp kp ledger in
+            Option.is_some account.zkapp )
+      with
+      | Some true ->
+          true
+      | _ ->
+          false
+
+    let all_zkapps_deployed ~ledger (keypairs : Signature_lib.Keypair.t list) =
+      List.map keypairs ~f:(fun kp -> is_zkapp_deployed kp ledger)
+      |> List.for_all ~f:Fn.id
+
+    let rec wait_until_zkapps_deployed ?(deployed = false) ~mina ~ledger
+        ~(fee_payer_keypair : Signature_lib.Keypair.t)
+        ~(account_creator_keypair : Signature_lib.Keypair.t)
+        ~constraint_constants ~logger ~uuid ~stop_signal ~stop_time
+        (keypairs : Signature_lib.Keypair.t list) =
+      if Time.( >= ) (Time.now ()) stop_time then (
+        [%log info] "Scheduled zkApp commands with handle %s has expired"
+          (Uuid.to_string uuid) ;
+        Uuid.Table.remove scheduler_tbl uuid ;
+        return None )
+      else if Ivar.is_full stop_signal then (
+        [%log info] "Stopping scheduled zkApp commands with handle %s"
+          (Uuid.to_string uuid) ;
+        Uuid.Table.remove scheduler_tbl uuid ;
+        return None )
+      else if all_zkapps_deployed ~ledger keypairs then (
+        [%log info] "All zkApp accounts are deployed" ;
+        return (Some ledger) )
+      else
+        let%bind () =
+          if not deployed then (
+            [%log info] "Start deploying zkApp accounts" ;
+            deploy_zkapps ~mina ~ledger ~fee_payer_keypair
+              ~account_creator_keypair ~constraint_constants ~logger keypairs )
+          else return ()
+        in
+        let%bind () =
```

Comment:
should this `after` be after the log?

should the log be `debug`, not sure we always need to see the entire ledger